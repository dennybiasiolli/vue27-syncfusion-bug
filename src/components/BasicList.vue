<template>
  <div>
    <ejs-grid
      ref="ejsGrid"
      :data-bound="handleDataBound"
      v-bind="gridParams"
      :action-complete="handleActionComplete"
      :edit-settings="computedEditSettings"
      :query-cell-info="handleQueryCellInfo"
      :row-selected="handleRowSelected"
      :toolbar="[]"
    >
      <e-columns>
        <e-column
          v-for="col in columns"
          :key="col.id"
          :foreign-key-field="col.foreignKeyField || 'id'"
          :foreign-key-value="col.foreignKeyValue || 'name'"
          v-bind="col"
        />
      </e-columns>
    </ejs-grid>
  </div>
</template>

<script>
import {
  DetailRow,
  Edit,
  Filter,
  ForeignKey,
  // Group,
  Page,
  // Reorder,
  Resize,
  Sort,
  Toolbar,
} from '@syncfusion/ej2-vue-grids';

export default {
  name: 'BasicList',
  provide: {
    grid: [
      DetailRow,
      Edit,
      Filter,
      ForeignKey,
      // Group,
      Page,
      // Reorder,
      Resize,
      Sort,
      Toolbar,
    ],
  },
  props: {
    columns: {
      type: Array,
      default: () => [],
    },
    gridParams: {
      type: Object,
      required: true,
    },
  },
  data() {
    return {
      innerEditSettings: {
        allowEditing: false,
        allowEditOnDblClick: false,
        allowAdding: false,
        allowDeleting: false,
        mode: 'Dialog',
        showDeleteConfirmDialog: true,
      },
    };
  },
  computed: {
    computedEditSettings() {
      return {
        ...this.gridParams?.editSettings,
        ...this.innerEditSettings,
      };
    },
  },
  methods: {
    handleActionComplete(e) {
      /* istanbul ignore else */
      if (e.requestType === 'add') {
        e.dialog.header = 'New Item';
      } else if (e.requestType === 'beginEdit') {
        e.dialog.header = 'Edit Item';
      }
      if (this.gridParams.actionComplete) {
        this.gridParams.actionComplete(e);
      }
    },
    handleDataBound() {
      if (this.gridParams.autoFitColumns) {
        this.$refs.ejsGrid.autoFitColumns();
      }
    },
    handleQueryCellInfo(e) {
      if (e.column.obfuscate) {
        this.$obfuscate.obfuscate(e.cell);
      }
      if (this.gridParams.queryCellInfo) {
        this.gridParams.queryCellInfo(e);
      }
    },
    handleRowSelected (e) {
      if (e.data.id) {
        this.innerEditSettings = {
          ...this.innerEditSettings,
          allowEditing: this.gridParams.editSettings?.allowEditing ?? e.data?.permissions?.change,
          allowEditOnDblClick: this.gridParams.editSettings?.allowEditOnDblClick ?? e.data?.permissions?.change,
          allowDeleting: this.gridParams.editSettings?.allowDeleting ?? e.data?.permissions?.delete,
        };
      }
      if (this.gridParams.rowSelected) {
        this.gridParams.rowSelected(e);
      }
    },
  },
};
</script>
