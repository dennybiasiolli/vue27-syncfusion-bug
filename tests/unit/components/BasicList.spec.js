import { shallowMount } from '@vue/test-utils';
import {
  DetailRow,
  Edit,
  Filter,
  ForeignKey,
  // Group,
  Page,
  // Reorder,
  Resize,
  Sort,
  Toolbar,
} from '@syncfusion/ej2-vue-grids';
import BasicList from '@/components/BasicList.vue';


describe('BasicList.vue', () => {
  const shallowMountComponent = params => {
    return shallowMount(BasicList, {
      ...params,
      stubs: ['ejs-grid', 'e-columns', 'e-column'],
    });
  };

  it('should provide tools to the grid', () => {
    expect(BasicList.provide).toEqual({
      grid: [
        DetailRow,
        Edit,
        Filter,
        ForeignKey,
        // Group,
        Page,
        // Reorder,
        Resize,
        Sort,
        Toolbar,
      ],
    });
  });

  it('should render as expected', async () => {
    const wrapper = shallowMountComponent({
      propsData: {
        columns: [{
          field: 'id',
          headerText: 'ID',
        }, {
          field: 'name',
          headerText: 'Name',
          foreignKeyField: 'foo',
          foreignKeyValue: 'bar',
        }],
        gridParams: {
          dataSource: [{ id: 1 }, { id: 2 }],
          toolbar: ['Add', 'Edit', 'Delete', 'Search', 'Print'],
        }
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should compute editSettings as expected', async () => {
    const editSettings = { allowEditing: true };
    const wrapper = shallowMountComponent({
      propsData: {
        columns: [],
        gridParams: {
          editSettings,
        },
      },
    });
    expect(wrapper.vm.innerEditSettings).toEqual({
      allowEditing: false,
      allowEditOnDblClick: false,
      allowAdding: false,
      allowDeleting: false,
      mode: 'Dialog',
      showDeleteConfirmDialog: true,
    });
    expect(wrapper.vm.computedEditSettings).toEqual({
      ...editSettings,
      ...wrapper.vm.computedEditSettings,
    });

    editSettings.allowDeleting = true;
    await wrapper.setProps({ editSettings });
    expect(wrapper.vm.computedEditSettings).toEqual({
      ...editSettings,
      ...wrapper.vm.computedEditSettings,
    });
  });

  it('should handleActionComplete correctly', async () => {
    const wrapper = shallowMountComponent({ propsData: { gridParams: {} } });
    const e = { dialog: {}, requestType: 'add' };
    wrapper.vm.handleActionComplete(e);
    expect(e.dialog.header).toBe('New Item');

    e.requestType = 'beginEdit';
    wrapper.vm.handleActionComplete(e);
    expect(e.dialog.header).toBe('Edit Item');


    const spyActionComplete = jest.fn();
    await wrapper.setProps({ gridParams: { actionComplete: spyActionComplete } });
    wrapper.vm.handleActionComplete(e);
    expect(spyActionComplete).toHaveBeenCalledWith(e);
  });

  it('should handleDataBound correctly', async () => {
    const wrapper = shallowMountComponent({ propsData: { gridParams: {} } });
    wrapper.vm.$refs.ejsGrid.autoFitColumns = jest.fn();

    wrapper.vm.handleDataBound();
    expect(wrapper.vm.$refs.ejsGrid.autoFitColumns).not.toHaveBeenCalled();

    await wrapper.setProps({ gridParams: { autoFitColumns: true } });
    wrapper.vm.handleDataBound();
    expect(wrapper.vm.$refs.ejsGrid.autoFitColumns).toHaveBeenCalledWith();
  });

  it('should handleQueryCellInfo correctly', async () => {
    const wrapper = shallowMountComponent({ propsData: { gridParams: {} } });
    wrapper.vm.$obfuscate = { obfuscate: jest.fn() };

    wrapper.vm.handleQueryCellInfo({ column: { obfuscate: false } });
    expect(wrapper.vm.$obfuscate.obfuscate).not.toHaveBeenCalled();

    const e = { cell: 'aaa', column: { obfuscate: true } };
    wrapper.vm.handleQueryCellInfo(e);
    expect(wrapper.vm.$obfuscate.obfuscate).toHaveBeenCalledWith(e.cell);


    const spyQueryCellInfo = jest.fn();
    await wrapper.setProps({ gridParams: { queryCellInfo: spyQueryCellInfo } });
    wrapper.vm.handleQueryCellInfo(e);
    expect(spyQueryCellInfo).toHaveBeenCalledWith(e);
  });

  it('should handleRowSelected correctly', async () => {
    const wrapper = shallowMountComponent({ propsData: { gridParams: {} } });
    let e = { data: {} };

    // disallow editing and deleting without explicit configuration
    wrapper.vm.handleRowSelected(e);
    expect(wrapper.vm.$data.innerEditSettings.allowEditing).toBe(false);
    expect(wrapper.vm.$data.innerEditSettings.allowEditOnDblClick).toBe(false);
    expect(wrapper.vm.$data.innerEditSettings.allowDeleting).toBe(false);

    // allow editing and deleting with explicit configuration in row data permissions
    e = { data: { id: 1, permissions: { change: true, delete: true } } };
    wrapper.vm.handleRowSelected(e);
    expect(wrapper.vm.$data.innerEditSettings.allowEditing).toBe(true);
    expect(wrapper.vm.$data.innerEditSettings.allowEditOnDblClick).toBe(true);
    expect(wrapper.vm.$data.innerEditSettings.allowDeleting).toBe(true);

    // disallow editing and deleting with explicit configuration in props
    await wrapper.setProps({ gridParams: { editSettings: { allowEditing: false, allowEditOnDblClick: false, allowDeleting: false } } });
    wrapper.vm.handleRowSelected(e);
    expect(wrapper.vm.$data.innerEditSettings.allowEditing).toBe(false);
    expect(wrapper.vm.$data.innerEditSettings.allowEditOnDblClick).toBe(false);
    expect(wrapper.vm.$data.innerEditSettings.allowDeleting).toBe(false);

    // allow editing with explicit configuration in props
    await wrapper.setProps({ gridParams: { editSettings: { allowEditing: true, allowEditOnDblClick: true, allowDeleting: false } } });
    wrapper.vm.handleRowSelected(e);
    expect(wrapper.vm.$data.innerEditSettings.allowEditing).toBe(true);
    expect(wrapper.vm.$data.innerEditSettings.allowEditOnDblClick).toBe(true);
    expect(wrapper.vm.$data.innerEditSettings.allowDeleting).toBe(false);

    const spyRowSelected = jest.fn();
    await wrapper.setProps({ gridParams: { rowSelected: spyRowSelected } });
    wrapper.vm.handleRowSelected(e);
    expect(spyRowSelected).toHaveBeenCalledWith(e);
  });
});
