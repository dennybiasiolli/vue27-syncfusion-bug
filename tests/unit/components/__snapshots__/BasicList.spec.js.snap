// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasicList.vue should render as expected 1`] = `
<div>
  <ejs-grid-stub action-complete="function () { [native code] }" data-bound="function () { [native code] }" datasource="[object Object],[object Object]" edit-settings="[object Object]" query-cell-info="function () { [native code] }" row-selected="function () { [native code] }" toolbar="Add,Edit,Delete,Search,Print">
    <e-columns-stub>
      <e-column-stub field="id" foreign-key-field="id" foreign-key-value="name" headertext="ID"></e-column-stub>
      <e-column-stub field="name" foreign-key-field="foo" foreign-key-value="bar" headertext="Name"></e-column-stub>
    </e-columns-stub>
  </ejs-grid-stub>
</div>
`;
